source/vstd/std_specs/hash.rs

```diff
@@ -0,0 +1,383 @@
+{
+    admit();
+}
+
+#[verifier::external_trait_specification]
+pub trait ExHasher {
+    type ExternalTraitSpecificationFor: Hasher;
+}
+
+// Our model for the external trait `BuildHasher` is that for any two
+// `Hasher`s it builds, if they're both given the same write sequence
+// then their states will match and they'll produce the same digest
+// when invoked with `finish()`.
+//
+// We don't expect that all types implementing the `BuildHasher` trait
+// will conform to our model, just the types T for which
+// `does_type_conform_to_build_hasher_model::<T>()` holds.
+#[verifier::external_trait_specification]
+pub trait ExBuildHasher {
+    type ExternalTraitSpecificationFor: BuildHasher;
+
+    type Hasher: Hasher;
+}
+
+#[verifier::external_body]
+pub spec fn does_type_conform_to_build_hasher_model<T: ?Sized>() -> bool;
+
+// A commonly used type of trait `BuildHasher` is `RandomState`. We
+// model that type here. In particular, we have an axiom that
+// `RandomState` conforms to our model of how `BuildHasher` behaves.
+#[verifier::external_type_specification]
+#[verifier::external_body]
+pub struct ExRandomState(RandomState);
+
+pub broadcast proof fn axiom_random_state_conforms_to_build_hasher_model()
+    ensures
+        does_type_conform_to_build_hasher_model::<RandomState>(),
+{
+    admit();
+}
+
+// We now specify the behavior of `HashMap`.
+#[verifier::external_type_specification]
+#[verifier::external_body]
+#[verifier::reject_recursive_types(Key)]
+#[verifier::reject_recursive_types(Value)]
+#[verifier::reject_recursive_types(S)]
+pub struct ExHashMap<Key, Value, S>(HashMap<Key, Value, S>);
+
+pub trait HashMapAdditionalSpecFns<Key, Value>: View<V = Map<Key, Value>> {
+    spec fn spec_index(&self, k: Key) -> Value
+        recommends
+            self@.contains_key(k),
+    ;
+}
+
+impl<Key, Value, S> HashMapAdditionalSpecFns<Key, Value> for HashMap<Key, Value, S> {
+    #[verifier::inline]
+    open spec fn spec_index(&self, k: Key) -> Value {
+        self@.index(k)
+    }
+}
+
+impl<Key, Value, S> View for HashMap<Key, Value, S> {
+    type V = Map<Key, Value>;
+
+    #[verifier::external_body]
+    closed spec fn view(&self) -> Map<Key, Value>;
+}
+
+pub open spec fn spec_hash_map_len<Key, Value, S>(m: &HashMap<Key, Value, S>) -> usize;
+
+pub broadcast proof fn axiom_spec_hash_map_len<Key, Value, S>(m: &HashMap<Key, Value, S>)
+    ensures
+        does_type_have_deterministic_hash::<Key>() && does_type_conform_to_build_hasher_model::<S>()
+            ==> #[trigger] spec_hash_map_len(m) == m@.len(),
+{
+    admit();
+}
+
+#[verifier::external_fn_specification]
+#[verifier::when_used_as_spec(spec_hash_map_len)]
+pub fn ex_hash_map_len<Key, Value, S>(m: &HashMap<Key, Value, S>) -> (len: usize)
+    ensures
+        len == spec_hash_map_len(m),
+{
+    m.len()
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_map_new<Key, Value>() -> (m: HashMap<Key, Value, RandomState>)
```

Comment:
Here and in `ex_hash_map_with_capacity`, the third generic argument to `HashMap` swaps from `S` (used above) to `RandomState`.  Is that intentional?

Reply:
Yes, this is intentional. The `new` function in the Rust standard library doesn't give you a choice of `S`. This is the literal signature from https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.new.

Reply:
Ah, interesting.  Thanks for clarifying!